File: Strivers-A2Z-DSA-Sheet/14. Dynamic Programming/8. DP on Partition/javascript/06. Partition array for maximum sum.js
/*
=============================================================================
  QUESTION: 1043. Partition Array for Maximum Sum (LeetCode)
=============================================================================

  Partition array into contiguous subarrays of length at most k.
  After partitioning, replace each element with the MAX of its subarray.
  Return maximum sum.

  Example: arr=[1,15,7,9,2,5,10], k=3 → 84
    Partition: [15,15,15] + [9,9] + [10,10] = 45+18+20 = 83? 
    Actually: [15,15] + [15] + [9,9] + [10,10] → 84

=============================================================================
  PATTERN: Front Partition DP
=============================================================================

  solve(i) = max sum for arr[i..n-1].
  At index i, try partition lengths 1 to k:
    For each j from i to min(i+k-1, n-1):
      subarray [i..j]: length = j-i+1, max element = maxi
      contribution = maxi × (j-i+1) + solve(j+1)
  Take maximum over all j.

  Base: i == n → 0

=============================================================================
  APPROACH: Memoization — O(N×K) Time, O(N) Space
=============================================================================
*/

function maxSumAfterPartitioning(arr, k) {
    let n = arr.length;
    let dp = new Array(n).fill(-1);

    function solve(i) {
        if (i === n) return 0;
        if (dp[i] !== -1) return dp[i];

        let maxi = 0, ans = 0;
        for (let j = i; j < Math.min(i + k, n); j++) {
            maxi = Math.max(maxi, arr[j]);
            // Replace all elements in [i..j] with maxi
            ans = Math.max(ans, maxi * (j - i + 1) + solve(j + 1));
        }
        return dp[i] = ans;
    }

    return solve(0);
}

// ==========================================================================
// DRIVER CODE
// ==========================================================================
console.log(maxSumAfterPartitioning([1, 15, 7, 9, 2, 5, 10], 3));  // 84

File: Strivers-A2Z-DSA-Sheet/14. Dynamic Programming/8. DP on Partition/python/06. Partition array for maximum sum.py
"""06. Partition Array for Max Sum. Subarrays ≤ k length."""
import sys; sys.setrecursionlimit(10000)
def max_sum_after_partitioning(arr, k):
    n = len(arr); dp = [-1]*n
    def solve(i):
        if i == n: return 0
        if dp[i] != -1: return dp[i]
        mx, ans = 0, 0
        for j in range(i, min(i+k, n)):
            mx = max(mx, arr[j])
            ans = max(ans, mx*(j-i+1) + solve(j+1))
        dp[i] = ans; return ans
    return solve(0)
if __name__ == "__main__": print(max_sum_after_partitioning([1,15,7,9,2,5,10], 3))  # 84

